veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,480 @@
+- A scoped abstraction keeps the project approachable for contributors while unlocking new hardware backends.
+
+## Goals
+
+- Document the cluster-wide hypervisor configuration and per-component extension points (defaults, converter, webhooks, node labeller) so downstream implementations can extend behavior without invasive changes to existing components.
+- Resolve the active hypervisor early and feed it through the virt-launcher converter so hypervisor-specific behavior stays localized.
+- Support both admission-time validation and mutation so administrators can enforce guardrails while still customizing VMIs for a given hypervisor.
+- Let components request allocatable device resources declared by the hypervisor configuration, avoiding new scheduling primitives.
+- Make it simple for downstreams to implement new hypervisors by following a documented contract.
+
+## Non Goals
+
+- Deliver a full implementation of any specific new hypervisor backend.
+- Redesign the VirtualMachineInstance API schema beyond additive fields.
+- Replace existing Hyper-V enlightenment features or other architecture-specific helpers.
+- Mandate new observability requirements; telemetry hooks remain optional.
+
+## Definition of Users
+
+- Cluster administrators who need to bootstrap KubeVirt on hardware that exposes alternative virtualization devices.
+- Platform vendors integrating proprietary or emerging hypervisor stacks with KubeVirt.
+- Upstream contributors maintaining virt-launcher, virt-controller, and virt-handler.
+
+## User Stories
+
+1. As a cluster administrator, I can declare a non-KVM hypervisor as the cluster default, and VMI pods schedule only on nodes that expose its required devices.
+2. As a platform engineer, I can supply hypervisor-specific VMI spec mutations and libvirt domain adjustments without forking the virt-launcher converter.
+3. As an upstream maintainer, I know exactly where to add validation, testing, and documentation when a new hypervisor is introduced.
+
+## Repos
+
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/enhancements](https://github.com/kubevirt/enhancements) (this VEP)
+
+## Design
+
+### Hypervisor Extension Points
+
+Cluster configuration (`spec.configuration.hypervisorConfiguration.name`) declares the active hypervisor for the entire installation, and each control-plane package exposes focused extension contracts so downstream implementations only touch the areas they actually need:
+
+- **Validation webhooks (`pkg/virt-api/webhooks/validating-webhook/admitters/hypervisor/`)** – We introduce a Validator interface that will define validation functions for core KubeVirt resources that have hypervisor-specific constraints, namely VM and VMI. Each hypervisor will provide its own concrete Validator to enforce rules and constraints relevant to its capabilities.
```

Comment:
Admission control webhooks can be a deployable integration point.
This takes us to the option of having this logic out-of-tree.

Reply:
What if the hypervisor is changed after a VM was admitted? Can this logic be moved into controllers instead?

Reply:
If we do so, the accelerator the VM first ran on would also need to be captured into its status to avoid changing it unknowingly. There more I think about it, the more I think we should keep multiple accelerator out of the API entirely for now.